src/aws_request.erl

```diff
@@ -0,0 +1,215 @@
+    AccessKeyID = aws_client:access_key_id(Client),
+    SecretAccessKey = aws_client:secret_access_key(Client),
+    Region = aws_client:region(Client),
+    Service = aws_client:service(Client),
+    Token = aws_client:token(Client),
+    Headers = case Token of
+                undefined -> Headers0;
+                _ -> [{<<"X-Amz-Security-Token">>, Token}|Headers0]
+              end,
+    aws_signature:sign_v4(AccessKeyID, SecretAccessKey, Region, Service, calendar:universal_time(), Method, URL, Headers, Body, Options).
+
+%% @doc Include additions only if they don't already exist in the provided list.
+add_headers([], Headers) ->
+  Headers;
+add_headers([{Name, _} = Header | Additions], Headers) ->
+  case lists:keyfind(Name, 1, Headers) of
+    false -> add_headers(Additions, [Header | Headers]);
+    _ -> add_headers(Additions, Headers)
+  end.
+
+%% @doc Build request headers based on a list key-value pairs
+%% representing the mappings from param names to header names and a
+%% map with the `params'.
+build_headers(ParamsHeadersMapping, Params0)
+  when is_list(ParamsHeadersMapping),
+       is_map(Params0) ->
+  Fun = fun({HeaderName, ParamName}, {HeadersAcc, ParamsAcc}) ->
+            case maps:get(ParamName, ParamsAcc, undefined) of
+              undefined ->
+                {HeadersAcc, ParamsAcc};
+              Value ->
+                Headers = [{HeaderName, Value} | HeadersAcc],
+                Params = maps:remove(ParamName, ParamsAcc),
+                {Headers, Params}
+            end
+        end,
+  lists:foldl(Fun, {[], Params0}, ParamsHeadersMapping).
+
+%% @doc Build custom request headers based on a list key-value pairs
+%% representing the mappings from param names to header names and a
+%% map with the `params'.
+build_custom_headers(ParamsCustomHeadersMapping, Params0)
+  when is_list(ParamsCustomHeadersMapping),
+       is_map(Params0) ->
+  Fun = fun({HeaderName, ParamName}, {HeadersAcc, ParamsAcc}) ->
+            case maps:get(ParamName, ParamsAcc, undefined) of
+              undefined ->
+                {HeadersAcc, ParamsAcc};
+              Value ->
+                Headers = [{<<HeaderName/binary, K/binary>>, V}
+                           || {K, V} <- maps:to_list(Value)] ++ HeadersAcc,
+                Params = maps:remove(ParamName, ParamsAcc),
+                {Headers, Params}
+            end
+        end,
+  lists:foldl(Fun, {[], Params0}, ParamsCustomHeadersMapping).
+
+%% @doc Add querystring to url is there are any parameters in the list
+-spec add_query(binary(), [{binary(), any()}]) -> binary().
+add_query(Url0, Query0) ->
+  UriMap0 = uri_string:parse(Url0),
+  ExistingQuery = uri_string:dissect_query(maps:get(query, UriMap0, [])),
+  NewQs = iolist_to_binary(aws_util:encode_query(ExistingQuery ++ Query0)),
+  UriMap =
+    case NewQs of
+      <<>> -> UriMap0;
+      _ -> UriMap0#{query => NewQs}
+    end,
+  uri_string:recompose(UriMap).
```

Comment:
This changed to use `uri_string` instead of `hackney_url` getting rid of it as a dependency for `aws_beam_core` 🎉  Added unit tests prior to changing to ensure compatibility between the two.